### PR TITLE
[Community pipeline] SD3 Differential Diffusion Img2Img Pipeline

### LINKED ISSUE #8577: Differential Diffusion with SD3
**Is your feature request related to a problem? Please describe.**
One might wish to soft inpaint or edit pictures with differential diffusion with SD3.

**Describe the solution you'd like.**
I believe we can port diff-diff to SD3, by adding a new pipeline for it.

**Describe alternatives you've considered.**
Diff-diff is already supported with XL.
Traditionally, diffusers supported inpainting via a custom checkpoint. 
I am not sure if SAI plans to release such a checkpoint for SD3.
 A port of differential diffusion for SD3 might fill this gap as well.

**Additional context.**
If I understand correctly, ComfyUI might already support diff-diff with SD3.

### PR DESCRIPTION
**Pull request by asomoza**
# What does this PR do?

Add the differential diffusion SD3 community pipeline.

Needs https://github.com/huggingface/diffusers/pull/8678 to work

Fixes #8577

How to test:

### Gradient
```python
import torch

from diffusers import FlowMatchEulerDiscreteScheduler
from diffusers.utils import load_image
from examples.community.pipeline_stable_diffusion_3_differential_img2img import (
    StableDiffusion3DifferentialImg2ImgPipeline,
)


pipe = StableDiffusion3DifferentialImg2ImgPipeline.from_pretrained(
    "stabilityai/stable-diffusion-3-medium-diffusers", torch_dtype=torch.float16, variant="fp16"
).to("cuda")

pipe.scheduler = FlowMatchEulerDiscreteScheduler.from_config(pipe.scheduler.config, shift=3.0)

prompt = "a green pear"

source_image = load_image(
    "https://huggingface.co/datasets/OzzyGT/testing-resources/resolve/main/differential/20240329211129_4024911930.png"
)
map = load_image(
    "https://huggingface.co/datasets/OzzyGT/testing-resources/resolve/main/differential/gradient_mask_2.png"
)

image = pipe(
    prompt=prompt,
    negative_prompt="",
    image=source_image,
    num_inference_steps=28,
    guidance_scale=4.5,
    strength=1.0,
    map=map,
).images[0]
```

|source|gradient|result|
|---|---|---|
||gradient_mask_2||

### Inpainting
```python
import torch

from diffusers import FlowMatchEulerDiscreteScheduler
from diffusers.utils import load_image
from examples.community.pipeline_stable_diffusion_3_differential_img2img import (
    StableDiffusion3DifferentialImg2ImgPipeline,
)


pipe = StableDiffusion3DifferentialImg2ImgPipeline.from_pretrained(
    "stabilityai/stable-diffusion-3-medium-diffusers", torch_dtype=torch.float16, variant="fp16"
).to("cuda")

pipe.scheduler = FlowMatchEulerDiscreteScheduler.from_config(pipe.scheduler.config, shift=3.0)

prompt = "Photorealistic close-up portrait of a Golden Retriever puppy. In the background, on the rug beside the puppy, a red ball, enticing the pup to play."
prompt_3 = "Photorealistic close-up portrait of a Golden Retriever puppy, around 4 months old, with a playful expression bathed in warm sunlight streaming through the window of a modern living room. Capture the soft texture of its fluffy fur, the glint of light in its big brown eyes, and a colorful bandana around its neck. In the background, on the rug beside the puppy, a red ball, enticing the pup to play."

source_image = load_image(
    "https://huggingface.co/datasets/OzzyGT/testing-resources/resolve/main/differential/dog_source.png"
)
mask = load_image(
    "https://huggingface.co/datasets/OzzyGT/testing-resources/resolve/main/differential/dog_inpainting_mask.png"
)

image = pipe(
    prompt=prompt,
    prompt_3=prompt_3,
    negative_prompt="",
    image=source_image,
    num_inference_steps=28,
    guidance_scale=4.5,
    strength=0.78,
    map=mask,
    max_sequence_length=512,
).images[0]

image.save("diff-diff-inpaint-result.png")
```

|source|mask|result|
|---|---|---|
|dog_source|dog_inpainting_mask|diff-diff-inpaint-result|

### Depthmap
```python
import torch

from diffusers import FlowMatchEulerDiscreteScheduler
from diffusers.utils import load_image
from examples.community.pipeline_stable_diffusion_3_differential_img2img import (
    StableDiffusion3DifferentialImg2ImgPipeline,
)


pipe = StableDiffusion3DifferentialImg2ImgPipeline.from_pretrained(
    "./models/stable_diffusion_3_medium",
    torch_dtype=torch.float16,
).to("cuda")

pipe.scheduler = FlowMatchEulerDiscreteScheduler.from_config(pipe.scheduler.config, shift=3.0)

prompt = "painting of a mountain landscape with a meadow and a forest, meadow background"

source_image = load_image(
    "https://huggingface.co/datasets/OzzyGT/testing-resources/resolve/main/differential/meadow.jpg"
)
depth_map = load_image(
    "https://huggingface.co/datasets/OzzyGT/testing-resources/resolve/main/differential/meadow_depth.png"
)

image = pipe(
    prompt=prompt,
    negative_prompt="",
    image=source_image,
    num_inference_steps=28,
    guidance_scale=4.5,
    strength=0.8,
    map=depth_map,
).images[0]

image.save("diff-diff-depth-result.png")
```

|source|depth map|result|
|---|---|---|
|meadow|meadow_depth|diff-diff-depth-result|

\* depth map made with marigold and diffusers

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

@yiyixuxu @vikm2o @exx8

### CODE REVIEW

**Comment by vikm2o** [comment]
@asomoza is `generator` required ?
When I run I still see the same issue 
```
--> 914     latents = original_with_noise[i] * mask + latents * (1 - mask)
    915 # end diff diff
    916 
    917 # expand the latents if we are doing classifier free guidance
    918 latent_model_input = torch.cat([latents] * 2) if self.do_classifier_free_guidance else latents

IndexError: index 1 is out of bounds for dimension 0 with size 1
```
I am using sd3-diff-diff branch.

**Comment by asomoza** [comment]
@vikm2o 

>  is generator required ?

No, that got in by mistake when I copied my code.

You need to use the `fix-flow-match-scale-noise` branch in my repo for it to work. They're in separate branches since I need to do separate PRs.

**Comment by vikm2o** [comment]
@asomoza Thanks, that worked!

File: examples/community/pipeline_stable_diffusion_3_differential_img2img.py [hunk 1]
```diff
@@ -0,0 +1,979 @@
...
+            raise ValueError(f"`max_sequence_length` cannot be greater than 512 but is {max_sequence_length}")
+
+    def get_timesteps(self, num_inference_steps, strength, device):
+        # get the original timestep using init_timestep
+        init_timestep = min(num_inference_steps * strength, num_inference_steps)
+
+        t_start = int(max(num_inference_steps - init_timestep, 0))
+        timesteps = self.scheduler.timesteps[t_start * self.scheduler.order :]
+        if hasattr(self.scheduler, "set_begin_index"):
+            self.scheduler.set_begin_index(t_start * self.scheduler.order)
```
Comment (yiyixuxu) [inline examples/community/pipeline_stable_diffusion_3_differential_img2img.py:599]:
```suggestion
```
related to https://github.com/huggingface/diffusers/pull/8678/files#r1651529021
let's not set a `begin_index` from diff-diff pipeline, this way we can treat the `scale_noise` for diff-diff similar to how it is used in the training, i.e. get the timestep index vis search

Comment (yiyixuxu) [reply examples/community/pipeline_stable_diffusion_3_differential_img2img.py:599]: with this change, we can then change the flow match`scale_noise` to have similar logic as https://github.com/huggingface/diffusers/blob/589931ca791deb8f896ee291ee481070755faa26/src/diffusers/schedulers/scheduling_euler_discrete.py#L618

Comment (asomoza) [reply examples/community/pipeline_stable_diffusion_3_differential_img2img.py:599]: how can I test when self.begin_index is not None?

With the current img2img self.begin_index is also None, so this change will make no difference with the code from before, or I'm missing something here?

Comment (yiyixuxu) [reply examples/community/pipeline_stable_diffusion_3_differential_img2img.py:599]: for img2img, it sets `begin_index` here, no? 

https://github.com/huggingface/diffusers/blob/5b51ad0052f587c55a7fa843fff9c6e7f3db0372/src/diffusers/pipelines/stable_diffusion_3/pipeline_stable_diffusion_3_img2img.py#L596

Comment (asomoza) [reply examples/community/pipeline_stable_diffusion_3_differential_img2img.py:599]: yeah, you're right that it's set there to 2, but I'm debugging it and when it gets to the scale_noise: 

image

it is `None`, I'll have to dig more why is this happening.

Comment (asomoza) [reply examples/community/pipeline_stable_diffusion_3_differential_img2img.py:599]: yeah, never mind, I was making a mistake. I was using `init_step_index`

**Review by yiyixuxu — APPROVED** [review]
ready to merge too?

**Comment by asomoza** [comment]
I'll try to remove the additional code from it before, I mean the need for torchvision and the need to preprocess the images outside of the pipeline.

**Comment by asomoza** [comment]
now is ready

**Comment by yiyixuxu** [comment]
merged! it is really nice!

**Pull request merged**